src/vsphere_cpi/spec/unit/cloud/vsphere/cloud_spec.rb

```diff
@@ -785,6 +788,267 @@ module VSphereCloud
+                    'pool_name' => 'fake-pool-2',
+                    'security_group' => fake_duplicate_sg,
+                    'port' => 443,
+                    'monitor_port' => 443,
+                  }
+                ]
+              }
+            }
+          end
+          let(:environment) do
+            {
+              'bosh' => {
+              }
+            }
+          end
+          it 'should de-duplicate the security tags and attach them to the VM' do
+            allow(VmConfig).to receive(:new).and_return(vm_config)
+            allow(NsxHttpClient).to receive(:new)
+              .with('fake-nsx-user', 'fake-nsx-password', 'fake-log-file')
+              .and_return(http_basic_auth_client)
+            allow(NSX).to receive(:new).and_return(nsx)
+
+            expect(VmCreator).to receive(:new).and_return(vm_creator)
+            expect(vm_creator).to receive(:create).with(vm_config).and_return(fake_vm)
+            expect(cloud_config).to receive(:validate_nsx_options)
+            expect(nsx).to receive(:add_vm_to_security_group).once
+            allow(nsx).to receive(:add_members_to_lbs)
+            vsphere_cloud.create_vm(
+              'fake-agent-id',
+              'fake-stemcell-cid',
+              vm_type,
+              'fake-networks-hash',
+              [],
+              environment
+            )
+          end
+        end
+
+        context 'when VM has duplicate security groups specified in bosh groups env but none in lbs and vm_type' do
+          let(:vm_type) do
+            {
+              'cpu' => 1,
+              'ram' => 1024,
+              'disk' => 4096,
+              'nsx' => {
+                'security_groups' => [],
+                'lbs' => []
+              }
+            }
+          end
+          let(:environment) do
+            {
+              'bosh' => {
+                'groups' => [
+                  fake_duplicate_sg,
+                  fake_duplicate_sg,
+                  another_fake_duplicate_sg,
+                  another_fake_duplicate_sg,
+                  another_fake_duplicate_sg,
+                  fake_unique_bosh_group,
+                  fake_unique_bosh_group,
+                ]
+              }
+            }
+          end
+          it 'should de-duplicate the security tags and attach them to the VM' do
+            allow(VmConfig).to receive(:new).and_return(vm_config)
+            allow(NsxHttpClient).to receive(:new)
+              .with('fake-nsx-user', 'fake-nsx-password', 'fake-log-file')
+              .and_return(http_basic_auth_client)
+            allow(NSX).to receive(:new).and_return(nsx)
+
+            expect(VmCreator).to receive(:new).and_return(vm_creator)
+            expect(vm_creator).to receive(:create).with(vm_config).and_return(fake_vm)
+            expect(cloud_config).to receive(:validate_nsx_options)
+            expect(nsx).to receive(:add_vm_to_security_group).exactly(3).times
```

Comment:
except this line almost all tests have same expectations, best is to refactor so that every 'it' block doesn't have any duplication.
Similarly for vm_type you could extract it out and each context can just set security_groups and lbs

Reply:
Can I take common expectation out and put them in a block? How?

If I extract vm_type out , I am only extracting ram, cpu out. Other things will still stay, so letting it be.